ddl/ddl_api.go

```diff
@@ -3383,200 +3383,138 @@ func checkModifyCharsetAndCollation(toCharset, toCollate, origCharset, origColla
-		}
-	case mysql.TypeTiny, mysql.TypeShort, mysql.TypeInt24, mysql.TypeLong, mysql.TypeLonglong:
-		switch to.Tp {
-		case mysql.TypeTiny, mysql.TypeShort, mysql.TypeInt24, mysql.TypeLong, mysql.TypeLonglong:
-			// Changing integer to integer, whether reorg is necessary is depend on the flen/decimal/signed.
-			skipSignCheck = true
-			skipLenCheck = true
-		default:
-			// Changing integer to other types, reorg is absolutely necessary.
-			return unsupportedMsg, errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
-		}
-	case mysql.TypeFloat, mysql.TypeDouble:
-		switch to.Tp {
-		case mysql.TypeFloat, mysql.TypeDouble:
-			skipSignCheck = true
-			skipLenCheck = true
-		case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp, mysql.TypeEnum, mysql.TypeSet:
-			// TODO: Currently float/double cast to date/datetime/timestamp/enum/set are all not support yet, should fix here after supported.
-			return "", errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
-		default:
-			return unsupportedMsg, errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
+// field length and precision. Any incompatible will return an error with individual error msg, and
+// if the "origin" type can be changed to "to" type by reorganization, the "canReorg" will be true,
+// false "canReorg" means the types change are not supported.
+func CheckModifyTypeCompatible(origin *types.FieldType, to *types.FieldType) (canReorg bool, errMsg string, err error) {
+	checkLengthDecimalSign := func() (bool, string, error) {
+		if to.Flen > 0 && to.Flen < origin.Flen {
+			msg := fmt.Sprintf("length %d is less than origin %d", to.Flen, origin.Flen)
+			return true, msg, errUnsupportedModifyColumn.GenWithStackByArgs(msg)
 		}
-	case mysql.TypeBit:
-		switch to.Tp {
-		case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp, mysql.TypeDuration, mysql.TypeEnum, mysql.TypeSet:
-			// TODO: Currently bit cast to date/datetime/timestamp/time/enum/set are all not support yet, should fix here after supported.
-			return "", errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
-		case mysql.TypeBit:
-		default:
-			return unsupportedMsg, errUnsupportedModifyColumn.GenWithStackByArgs(unsupportedMsg)
+		if to.Decimal > 0 && to.Decimal < origin.Decimal {
+			msg := fmt.Sprintf("decimal %d is less than origin %d", to.Decimal, origin.Decimal)
+			return true, msg, errUnsupportedModifyColumn.GenWithStackByArgs(msg)
 		}
-	case mysql.TypeEnum, mysql.TypeSet:
-		var typeVar string
-		if origin.Tp == mysql.TypeEnum {
-			typeVar = "enum"
-		} else {
-			typeVar = "set"
+		if mysql.HasUnsignedFlag(origin.Flag) != mysql.HasUnsignedFlag(to.Flag) {
+			msg := fmt.Sprintf("can't change unsigned integer to signed or vice versa")
+			return true, msg, errUnsupportedModifyColumn.GenWithStackByArgs(msg)
 		}
-		switch to.Tp {
-		case mysql.TypeEnum, mysql.TypeSet:
+		return false, "", nil
```

Comment:
`false` with `nil` error, it seems weird.

Reply:
Agreed. Maybe I can extract the assemble error logic out of scope, so the function can only need to return `bool, string`.